# **Numpy**

In [1]:
import numpy as np

Essa lista contém questões práticas incluindo os seguintes tópicos com numpy.

* Criação de array e seus atributos, intervalos numéricos em numPy
* Slicing e indexação de NumPy Array.
* Manipulação, pesquisa, classificação e divisão de arrays.
* Funções matemáticas com arrays e broadcasting arrays NumPy. 

## **Crie um array inteiro 4x2 e mostre seus atributos**

Os elementos deven ser um tipo de int16 sem sinal. E imprima os seguintes atributos: formato do array; dimensões; comprimento de cada elemento do array.

In [4]:
#########################################
arr = np.array([
    [64392, 31655],
    [32579, 0],
    [49248, 462],
    [0, 0]
], dtype=np.uint16)

print('Mostrando o array:')
print(arr)
print('Mostrando os atributos:')
print('Formato do array:',arr.shape)
print('Número de dimensões:',arr.ndim)
print('Comprimento de cada elemento do array:',arr.itemsize)
#########################################

Mostrando o array:
[[64392 31655]
 [32579     0]
 [49248   462]
 [    0     0]]
Mostrando os atributos:
Formato do array: (4, 2)
Número de dimensões: 2
Comprimento de cada elemento do array: 2


##### Resultado esperado:
<pre>
Mostrando Array
[[64392 31655]
 [32579     0]
 [49248   462]
 [    0     0]]

Mostrando Attributos:

Formato do array:  (4, 2)
Número de dimensões:  2
Comprimento de cada elemento do array:  2
</pre>

## **Crie uma array de inteiros 5X2 com um intervalo entre 100 e 200, de modo que a diferença entre cada elemento seja 10**

In [12]:
#########################################
arr = np.arange(start=100, stop=200, step=10).reshape((5,2))
print(arr)
#########################################

[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]


##### Resultado esperado:
<pre>
[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]
</pre>

## **Baseado no array fornecido. Retorne um array de itens com a terceira coluna de todas as linhas**

In [13]:
array_00 = np.array([[11 ,22, 33], [44, 55, 66], [77, 88, 99]])

In [16]:
#########################################
arr = array_00[:,2]

print('Mostrando array de entrada:')
print(array_00)
print('Novo array com a terceira coluna de cada linha:')
print(arr)
#########################################

Mostrando array de entrada:
[[11 22 33]
 [44 55 66]
 [77 88 99]]
Novo array com a terceira coluna de cada linha:
[33 66 99]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[11 22 33]
 [44 55 66]
 [77 88 99]]

Novo array com a terceira coluna de cada linha
[33 66 99]
</pre>

## **Retorne um array com os valores presentes nas linhas ímpares com colunas pares baseado no seguinte array**

In [18]:
array_01 = np.array([[3 ,6, 9, 12], [15 ,18, 21, 24], 
[27 ,30, 33, 36], [39 ,42, 45, 48], [51 ,54, 57, 60]])

In [22]:
#########################################
arr = array_01[1::2,0::2]
print('Mostrano array de entrada:')
print(array_01)
print('Array resultante (linha ímpar x coluna par):')
print(arr)
#########################################

Mostrano array de entrada:
[[ 3  6  9 12]
 [15 18 21 24]
 [27 30 33 36]
 [39 42 45 48]
 [51 54 57 60]]
Array resultante (linha ímpar x coluna par):
[[15 21]
 [39 45]]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[ 3  6  9 12]
 [15 18 21 24]
 [27 30 33 36]
 [39 42 45 48]
 [51 54 57 60]]

Array resultante (linha ímpar x coluna par)
[[ 6 12]
 [30 36]
 [54 60]]
</pre>

## **Crie um array de resultado somando os seguintes arrays. Em seguida, modifique o array resultante calculando o quadrado de cada elemento**

In [24]:
array_02 = np.array([[5, 6, 9], [21 ,18, 27]])
array_03 = np.array([[15 ,33, 24], [4 ,7, 1]])

In [26]:
#########################################
arr = array_02 + array_03
arr = np.power(arr, 2)
#########################################

##### Resultado esperado:
<pre>
A soma dos arrays é: 
[[20 39 33]
 [25 25 28]]

Resultado depois de calcular o quadrado de cada número:
[[ 400 1521 1089]
 [ 625  625  784]]
</pre>

## **Faça a soma, subtração, multiplicação e divisão dos arrays**

In [27]:
array_04 = np.array([1, 2, 3])
array_05 = np.array([[1, 2, 3], [4, 5, 6]])

In [28]:
#########################################
soma = array_04 + array_05
sub = array_04 - array_05
mult = array_04 * array_05
div = array_04 / array_05
print('Soma:')
print(soma)
print('Subtração:')
print(sub)
print('Multiplicação:')
print(mult)
print('Divisão:')
print(div)
#########################################

Soma:
[[2 4 6]
 [5 7 9]]
Subtração:
[[ 0  0  0]
 [-3 -3 -3]]
Multiplicação:
[[ 1  4  9]
 [ 4 10 18]]
Divisão:
[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]


##### Resultado esperado:
<pre>
Soma: 
[[2 4 6]
 [5 7 9]]
Subtração: 
[[ 0  0  0]
 [-3 -3 -3]]
Multiplicação: 
[[ 1  4  9]
 [ 4 10 18]]
divisão: 
[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]
</pre>

## **Divida o array em quatro sub-arrays de tamanhos iguais**

Crie um array de inteiros 8X3 com um intervalo entre 10 e 34 de modo que a diferença entre cada elemento seja 1 e, em seguida, faça a subdivisão. 

In [35]:
#########################################
arr = np.arange(start=10, stop=34, step=1).reshape((8,3)).astype(np.uint32)
print('Criando um array 8x3 com o arange:')
print(arr)
sub_arrs = np.array_split(arr, 4)
print('Dividindo o array em 4 subarrays:')
print(sub_arrs)
#########################################

Criando um array 8x3 com o arange:
[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]
Dividindo o array em 4 subarrays:
[array([[10, 11, 12],
       [13, 14, 15]], dtype=uint32), array([[16, 17, 18],
       [19, 20, 21]], dtype=uint32), array([[22, 23, 24],
       [25, 26, 27]], dtype=uint32), array([[28, 29, 30],
       [31, 32, 33]], dtype=uint32)]


##### Resultado esperado:
<pre>
Criando o array de 8X3 usando numpy.arange
[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]

Dividindo o array em 4 sub-arrays

[array([[10, 11, 12],[13, 14, 15]]), 
array([[16, 17, 18],[19, 20, 21]]), 
array([[22, 23, 24],[25, 26, 27]]), 
array([[28, 29, 30],[31, 32, 33]])]
</pre>

## **Ordenar o seguinte array da seguinte forma:**

* Oredene pela segunda linha
* Ordene pela terceira coluna

In [47]:
import numpy as np
array_06 = np.array([[34,43,73],[82,22,12],[53,94,66]])

In [48]:
array_06

array([[34, 43, 73],
       [82, 22, 12],
       [53, 94, 66]])

In [55]:
arr = array_06.copy()
arr[1, :].sort()
print('Ordenando o array pela segunda linha:')
print(arr)
arr = array_06.copy()
arr[:, 2].sort()
print('Ordenando pela terceira coluna:')
print(arr)

Ordenando o array pela segunda linha:
[[34 43 73]
 [12 22 82]
 [53 94 66]]
Ordenando pela terceira coluna:
[[34 43 12]
 [82 22 66]
 [53 94 73]]


##### Resultado esperado:
<pre>
Mostrando o array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Ordenando o array pela segunda linha
[[73 43 34]
 [12 22 82]
 [66 94 53]]

Ordenando o array pela terceira coluna
[[82 22 12]
 [34 43 73]
 [53 94 66]]
</pre>

## **Imprima o máx. do eixo 0 e mín. Do eixo 1 do seguinte array**

In [37]:
array_07 = np.array([[34,43,73],[82,22,12],[53,94,66]])

In [45]:
#########################################
print('Max no eixo 0:')
print(np.max(array_07,axis=0))
print('Min no eixo 1:')
print(np.min(array_07,axis=1))
#########################################

Max no eixo 0:
[82 94 73]
Min no eixo 1:
[34 12 53]


##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Mostando amin do eixo 1
[34 12 53]

Mostrando amax do eixo 0
[82 94 73]
</pre>

## **Exclua a segunda coluna de um determinado array e insira a nova coluna seguinte em seu lugar**

In [41]:
array_08    = np.array([[34,43,73],[82,22,12],[53,94,66]]) 
nova_coluna = np.array([[10,10,10]])

In [44]:
#########################################
print('Array Original:')
print(array_08)
arr = np.delete(array_08, 1, axis=1)
print('Array sem a segunda coluna:')
print(arr)
print('Array com a nova segunda coluna:')
arr = np.insert(arr, 1, nova_coluna, axis=1)
print(arr)
#########################################

Array Original:
[[34 43 73]
 [82 22 12]
 [53 94 66]]
Array sem a segunda coluna:
[[34 73]
 [82 12]
 [53 66]]
Array com a nova segunda coluna:
[[34 10 73]
 [82 10 12]
 [53 10 66]]


##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Array depois de deletar a segunda coluna
[[34 73]
 [82 12]
 [53 66]]

Array depois de inserir a nova coluna no lugar da coluna excluída
[[34 10 73]
 [82 10 12]
 [53 10 66]]
</pre>